### ISSUE TITLE
Support PPO with value model, not value head

### ISSUE THREAD

**Issue by natolambert** [huggingface/trl#1063]
There's a lot of effort brewing to do thorough comparisons of DPO to PPO and other methods, but to the best of my knowledge many of the best PPO successes in public (and mostly private) use entire value models instead of just value heads. 
Any chance we can have help introducing this? Anyone? 🤗

**Comment by natolambert** [comment]
I think it'll be an additional option / parameter (non-default) then everything needs to be validated again. It's tough, but worth it?

The value network is interesting. Really odd, not sure why it's so bad. 
Maybe the normalization is more important with a value model rather than a value head, given it has less regularization by "More different" parameters?

**Comment by natolambert** [comment]
Related, @vwxyzjn shared this photo from the appendix of the summarize work with me 
![image](https://github.com/huggingface/trl/assets/10695622/40777242-9b8f-443e-8778-31258731bc5f)

**Comment by vwxyzjn** [comment]
Adding a note here:

To use separate value head for PPO, we might not be able to use AutoModelForSequenceClassification without refactoring transformers.

The reason is that AutoModelForSequenceClassification only returns the pooled_logits corresponding to EOS tokens ([modeling_gpt_neox.py#L1214](https://github.com/huggingface/transformers/blob/45b1dfa342f2ec00634c31f635bded6e346c98d2/src/transformers/models/gpt_neox/modeling_gpt_neox.py#L1214)), there isn't any way to get the actual logits corresponding to each token, but for separate value network we need to model the value at each token.

**Comment by vwxyzjn** [comment]
A current workaround is something like 


```

class ScalarModelConfig(PretrainedConfig):
    def __init__(
        self,
        base_model: str = "EleutherAI/pythia-160m",
        base_config: PretrainedConfig = AutoConfig.from_pretrained("EleutherAI/pythia-160m"),
        hidden_size: int = 768,
        bias: float = 0.0,
        **kwargs,
    ):
        super().__init__(**kwargs)
        self.base_model = base_model
        self.base_config = base_config
        self.hidden_size = hidden_size
        self.bias = bias


class ScalarModel(PreTrainedModel):
    config_class = ScalarModelConfig

    def __init__(self, config: ScalarModelConfig):
        super().__init__(config)
        self.config = config
        self.lm_backbone = AutoModel.from_pretrained(
            config.base_model,
            config=self.config.base_config,
            trust_remote_code=True,
        )
        self.scalar_head = layer_init(
            nn.Linear(self.config.hidden_size, 1),
            std=1 / np.sqrt(self.config.hidden_size + 1),
        )

    def forward(self, **kwargs):
        output = self.lm_backbone(**kwargs)
        reward = self.scalar_head(output.hidden_states[-1]) - self.config.bias
        return reward
```

**Comment by Atry** [comment]
> Related, @vwxyzjn shared this photo from the appendix of the summarize work with me ![image](https://private-user-images.githubusercontent.com/10695622/294020114-40777242-9b8f-443e-8778-31258731bc5f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5MjI4ODUsIm5iZiI6MTcxODkyMjU4NSwicGF0aCI6Ii8xMDY5NTYyMi8yOTQwMjAxMTQtNDA3NzcyNDItOWI4Zi00NDNlLTg3NzgtMzEyNTg3MzFiYzVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIwVDIyMjk0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU1NzBiYzE3ZmVhZDRhMzhmOTRiMDE0MzFmYTY0MzBhMDBhMDNhYmU5NmUzMjMxMDU4ZTI4NGM0MTQxOWFiODUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.HfQafDSaa79RAwKqSu5hYjbMWyuGZQeTsu81nh3eIvg)

Source: https://openai.com/index/learning-to-summarize-with-human-feedback/
